cdc/generation.cc

```diff
@@ -328,4 +356,290 @@ void update_streams_description(
+                cdc_log.info("Generation {} already known.", ts);
+                return;
+            }
+
+            cdc_log.info("Learned generation {}. Updating system_distributed.cdc_description...", ts);
+
+            while (true) {
+                try {
+                    do_update_streams_description(
+                            ts, *get_sys_dist_ks(_sys_dist_ks), { _ring.count_normal_token_owners() });
+                    return;
+                } catch (...) {
+                    cdc_log.warn(
+                        "Failed to update system_distributed.cdc_description with generation {}."
+                        " Reason: \"{}\". Retrying...",
+                        ts, std::current_exception());
+
+                    sleep_abortable(std::chrono::seconds(60), _abort_source).get();
+                }
+            }
+        });
+    }).handle_exception([ts] (std::exception_ptr ep) {
+        // The only possible exception here should be coming from sleep_abortable or the gate.
+        // This means we're shutting down. No need to do anything fancy.
+        try {
+            std::rethrow_exception(ep);
+        } catch (...) {
+            cdc_log.warn("Error while learning generation {}: {}", ts, std::current_exception());
+        }
+    });
+}
+
+void generation_service::new_generation(
+        const std::unordered_set<dht::token>& bootstrap_tokens,
+        bool in_single_node_test) {
+    using namespace std::chrono;
+    assert_shard_zero(__PRETTY_FUNCTION__);
+
+    cdc_log.info("Creating a new CDC generation...");
+
+    auto desc = topology_description_generator(_db_cfg, bootstrap_tokens, _ring, _gossiper).generate();
+
+    // For propagating our generation, we will depend on the (not always true) ring_delay assumption:
+    // anything inserted into gossip will arrive at other live nodes within ring_delay.
+    // We're adding 2 * ring_delay to give us just a little more time.
+    // If we're currently running a single-node unit test it's not necessary to add any delay.
+    auto ts = db_clock::now() + (
+            in_single_node_test
+                ? milliseconds(0)
+                : (2 * milliseconds(_db_cfg.ring_delay_ms()) + duration_cast<milliseconds>(generation_leeway)));
+
+    cdc_log.info("New generation's timestamp: {}", ts);
+
+    get_sys_dist_ks(_sys_dist_ks)->insert_cdc_topology_description(
+            ts, std::move(desc), { _ring.count_normal_token_owners() }).get();
+
+    cdc_log.info("Inserted generation {} into system_distributed.cdc_topology_description.", ts);
+
+    update_generation_timestamp(ts);
+}
+
+void generation_service::before_join_token_ring(
+        const std::unordered_set<dht::token>& bootstrap_tokens,
+        const gms::feature_service& feature_svc,
+        bool in_single_node_test) {
+    assert_shard_zero(__PRETTY_FUNCTION__);
+    assert(!db::system_keyspace::bootstrap_complete());
+
+    if (!_db_cfg.check_experimental(db::experimental_features_t::CDC)) {
```

Comment:
Isn't this already handled by features mechanism?

Reply:
I don't understand the question.
We need to check if CDC is enabled on this node before creating a new CDC generation.
If it's disabled, we don't create.

We could instead check if the whole cluster supports CDC but that's not necessary. The first node which enables CDC on rolling upgrade can already safely create a CDC generation: other nodes will just ignore it until they upgrade.

Reply:
Makes sense.